src/RadixDlt.NetworkGateway.PostgresIntegration/Models/EntityResourceVaultAggregateHistory.cs

```diff
@@ -84,6 +89,107 @@ internal class EntityResourceVaultAggregateHistory
     [Column("resource_entity_id")]
     public long ResourceEntityId { get; set; }
 
+    /// <summary>
+    /// The aggregate list of all vaults ordered by most recently updated.
+    /// </summary>
     [Column("vault_entity_ids")]
     public List<long> VaultEntityIds { get; set; }
+
+    /// <summary>
+    /// The last significant update state version for resources stored in <see cref="VaultEntityIds"/>.
+    /// </summary>
+    /// <remarks>
+    /// This property is nullable in order to ensure smooth database migration process.
+    /// </remarks>
+    [Column("vault_significant_update_state_versions")]
+    public List<long>? VaultSignificantUpdateStateVersions { get; set; }
+
+    public static EntityResourceVaultAggregateHistory Create(long databaseId, long entityId, long resourceEntityId, long stateVersion)
+    {
+        return CopyOrCreate(databaseId, null, entityId, resourceEntityId, stateVersion);
+    }
+
+    public static EntityResourceVaultAggregateHistory CopyOf(long databaseId, EntityResourceVaultAggregateHistory other, long stateVersion)
+    {
+        return CopyOrCreate(databaseId, other, other.EntityId, other.ResourceEntityId, stateVersion);
+    }
+
+    /// <summary>
+    /// Attempts to add new or update existing vault unless it already is the most recently modified one.
+    /// </summary>
+    /// <returns>true if added or modified, otherwise false.</returns>
+    public bool TryUpsert(long vaultId, long stateVersion)
+    {
+        EnsureVaultSignificantUpdateStateVersionsExists();
+
+        var currentIndex = VaultEntityIds.IndexOf(vaultId);
+
+        // we're already the most recent one and there's no point potential second resource with the same state version
+        if (currentIndex == 0 && VaultEntityIds.Count == 1)
```

Comment:
These two if statements feel a bit arbitrary / overly optimized, and mean that `VaultSignificantUpdateStateVersions` isn't an accurate name - we don't update the state version of an item at the top of the list. Personally, I'd rather we just removed them both.

If we call it `__StateVersions`, I'd rather it was _accurately_ the state version of the last update. Otherwise, we should call it `RelativeUpdateOrdering` or something.